samcli/hook_packages/terraform/designs/resource_linking_generalized.md

```diff
@@ -0,0 +1,153 @@
+
+Once all `ResourceLinkingPair` objects have been created, a list of these objects can be passed to the `ResourceLinker`.
+```python
+class ResourceLinker:
+    _resource_pairs: List[ResourceLinkingPair]
+
+    def link_resources(self):
+        """
+        Iterate through all of the ResourceLinkingPair items and link the
+        corresponding source resource to destination resource
+        """
+
+    def _update_mapped_parent_resource_with_resolved_child_resources(self, destination_resources: List):
+        """
+        Set the resolved destination resource list to the mapped source resources.
+
+        Parameters
+        ----------
+        destination_resources: List
+            The resolved destination resource values that will be used as a value for the mapped CFN resource attribute.
+        """
+
+    def _process_reference_resource_value(self, resolved_destination_resource: ResolvedReference):
+        """
+        Process the a reference destination resource value of type ResolvedReference.
+
+        Parameters
+        ----------
+        resolved_destination_resource: ResolvedReference
+            The resolved destination resource reference.
+
+        Returns
+        -------
+        List[Dict[str, str]]
+            The resolved values that will be used as a value for the mapped CFN resource attribute.
+        """
+
+    def _process_resolved_resources(self, resolved_destination_resource: List[Union[ConstantValue, ResolvedReference]]):
+        """
+        Process the resolved destination resources.
+
+        Parameters
+        ----------
+        resolved_destination_resource: List[Union[ConstantValue, ResolvedReference]]
+            The resolved destination resources to be processed for the input source resource.
+
+        Returns
+        --------
+        List[Dict[str, str]]:
+            The list of destination resources after processing
+        """
+```
+
+The `ResourceLinker` contains a public `link_resources` method to begin execution of the resource linking. The methods 
+shown in the class are all generalized versions of the existing Function to Layer linking functions that exist today.
+
+`link_resources` &larr; `_link_lambda_function_to_layer`, `_link_lambda_functions_to_layers`
+`_update_mapped_parent_resource_with_resolved_child_resources` &larr; `_update_mapped_lambda_function_with_resolved_layers`
+`_process_reference_resource_value` &larr; `_process_reference_layer_value`
+`_process_resolved_resources` &larr; `_process_resolved_layers` 
+
+To generalize these functions: 
+- All instances of hard-coded properties need to be replaced with the corresponding values
+defined in the `ResourceLinkingPair` instance
+- Exception messages should be updated to be more generic
+- The `_update_mapped_parent_resource_with_resolved_child_resources` method should be updated to support writing other intrinsic types.
+
+
+### Adding New Resource Links
```

Comment:
we need also to change the linking exceptions we throw (mainly the limitations exceptions) for each linking pair. I think it can be done by adding a new Exception Type, and in the new `ResourceLinkingPair` instance we can set the Exceptions type to be used.

Reply:
The only Function/Layer specific exception I can see is `OneLambdaLayerLinkingLimitationException`. Are there any others?